File: expensereport-elixir/expensereport.exs
#!/usr/bin/env elixir

defmodule Expense do
    defstruct [:type, :amount]
end

defmodule ExpenseReport do
    def printReport(htmlMode, expenses) when is_list(expenses) do

        if htmlMode do
            IO.write("""
                <!DOCTYPE html>
                <html lang="en">
                <head>
                <title>Expense Report</title>
                </head>
                <body>
                <h1>Expense Report</h1>
                """)
        else
            IO.puts("Expense Report")
        end

        if htmlMode do
            IO.write("""
                <table>
                <thead>
                <tr><th scope="col">Type</th><th scope="col">Amount</th><th scope="col">Over Limit</th></tr>
                </thead>
                <tbody>
                """)
        end
        for expense <- expenses do
            expenseName = case expense.type do
                :breakfast -> "Breakfast"
                :dinner -> "Dinner"
                :carrental -> "Car Rental"
            end
            mealOverExpensesMarker = case expense.type do
                :breakfast -> if expense.amount > 1000 do "X" else " " end
                :dinner -> if expense.amount > 5000 do "X" else " " end
                _ -> " "
            end
            if htmlMode do
                IO.puts("<tr><td>" <> expenseName <> "</td><td>" <> to_string(expense.amount) <> "</td><td>" <> mealOverExpensesMarker <> "</td></tr>")
            else
                IO.puts(expenseName <> "\t" <> to_string(expense.amount) <> "\t" <> mealOverExpensesMarker)
            end
        end
        if htmlMode do
            IO.write("""
                </tbody>
                </table>
                """)
        end

        mealExpenses = Enum.sum(Enum.map(Enum.filter(expenses, fn expense -> expense.type == :breakfast || expense.type == :dinner end), fn expense -> expense.amount end))
        total = Enum.sum(Enum.map(expenses, fn expense -> expense.amount end))

        if htmlMode do
            IO.puts("<p>Meal expenses: " <> to_string(mealExpenses) <> "</p>")
            IO.puts("<p>Total expenses: " <> to_string(total) <> "</p>")
        else
            IO.puts("Meal expenses: " <> to_string(mealExpenses))
            IO.puts("Total expenses: " <> to_string(total))
        end

        if htmlMode do
            IO.write("""
                </body>
                </html>
                """)
        end
    end
end

defmodule Main do
    def main do
        ExpenseReport.printReport(true, [
            %Expense{type: :breakfast, amount: 1000},
            %Expense{type: :breakfast, amount: 1001},
            %Expense{type: :dinner, amount: 5000},
            %Expense{type: :dinner, amount: 5001},
            %Expense{type: :carrental, amount: 4},
        ])
    end
end

Main.main

File: expensereport-ruby/ExpenseReport.rb
#!/usr/bin/ruby

class Expense
  attr_reader :type, :amount
  def initialize(type, amount)
    @type = type
    @amount = amount
  end
end

def printReport(html_mode, *expenses)
  total = 0
  mealExpenses = 0

  if html_mode
    puts <<END
<!DOCTYPE html>
<html lang="en">
<head>
<title>Expense Report #{Time.now}</title>
</head>
<body>
<h1>Expense Report #{Time.now}</h1>
END
  else
    puts "Expenses: #{Time.now}"
  end

  if html_mode
    puts <<END
<table>
<thead>
<tr><th scope="col">Tyep</th><th scope="col">Amount</th><th scope="col">Over Limit</th></tr>
</thead>
<tbody>
END
  end
  for expense in expenses
    if expense.type == :dinner || expense.type == :breakfast
      mealExpenses += expense.amount
    end
    expenseName = ""
    case expense.type
    when :breakfast
        expenseName = "Breakfast"
    when :dinner
        expenseName = "Dinner"
    when :car_rental
        expenseName = "Car Rental"
    end
    mealOverExpensesMarker = expense.type == :dinner && expense.amount > 5000 || expense.type == :breakfast && expense.amount > 1000 ? "X" : " "
    if html_mode
      puts "<tr><td>#{expenseName}</td><td>#{expense.amount}</td><td>#{mealOverExpensesMarker}</td></tr>"
    else
      puts "#{expenseName}\t#{expense.amount}\t#{mealOverExpensesMarker}"
    end
    total += expense.amount
  end
  if html_mode
    puts "</tbody>"
    puts "</table>"
  end

  if html_mode
    puts "<p>Meal Expenses: #{mealExpenses}</p>"
    puts "<p>Total Expenses: #{total}</p>"
  else
    puts "Meal Expenses: #{mealExpenses}"
    puts "Total Expenses: #{total}"
  end

  if html_mode
    puts "</body>"
    puts "</html>"
  end
end

printReport(
    true,
    Expense.new(:breakfast, 1000),
    Expense.new(:breakfast, 1001),
    Expense.new(:dinner, 5000),
    Expense.new(:dinner, 5001),
    Expense.new(:car_rental, 4),
)
